### PR TITLE
Support struct in template (clean history)

### PR DESCRIPTION
**Pull request by ansharlubis**
Added a simple example with struct in template. To handle this, I moved the template instantiation to symbol table visitor.

### CODE REVIEW

File: integration_tests/template_array_02b.f90 [hunk 1]
```diff
@@ -62,10 +62,10 @@ function mysum_t(n, a) result(r)
             integer, intent(in) :: n
             type(t), intent(in) :: a(n)
             type(t) :: r
-            integer :: i
+            integer :: j
             r = zero_t(a(1))
-            do i = 1, size(a)
-                r = plus_t(r, a(i))
+            do j = 1, size(a)
+                r = plus_t(r, a(j))
```
Comment (certik) [inline integration_tests/template_array_02b.f90:68]: Why is `i` renamed to `j`?

**Comment by certik** [comment]
Thanks! There is a lot of code, once this is ready, ping me and I'll review carefully. Thanks for working on this.

File: integration_tests/template_array_02b.f90 (hunk 1 above)
Comment (ansharlubis) [reply integration_tests/template_array_02b.f90:68]: I found some scoping problems with the variable once I moved instantiation to symbol table visitor, but I haven't pinpointed exactly the reason.

Comment (ansharlubis) [reply integration_tests/template_array_02b.f90]: I tried reverting the example back, and somehow they are working properly now.

**Comment by ansharlubis** [comment]
@certik There are some problems with the build checks which I am not familiar with, but the code should be ready for review now.

File: integration_tests/template_add_04.f90 [hunk 2]
```diff
@@ -0,0 +1,60 @@
+module template_add_01_m
```
Comment (certik) [inline integration_tests/template_add_04.f90:1]: The error at the CI:
```
Traceback (most recent call last):
FAIL: Found a duplicate module name
  File "/home/runner/work/lfortran/lfortran/integration_tests/./run_tests.py", line 101, in <module>
Name: template_add_01_m
    main()
Filename: template_add_04.f90
  File "/home/runner/work/lfortran/lfortran/integration_tests/./run_tests.py", line 85, in main
    check_module_names()
  File "/home/runner/work/lfortran/lfortran/integration_tests/./run_tests.py", line 65, in check_module_names
    raise Exception("Duplicate module names")
Exception: Duplicate module names
```
Means you have a duplicate module name. To fix it:
```suggestion
module template_add_04_m
```

**Comment by certik** [comment]
You will need to update reference tests as well.

**Comment by certik** [comment]
The Windows build error seems real:
```
   Creating library src\bin\lfortran.lib and object src\bin\lfortran.exp
lfortran_lib.lib(ast_body_visitor.cpp.obj) : error LNK2019: unresolved external symbol "void __cdecl LCompilers::pass_instantiate_function_body(class Allocator &,class std::map<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,struct LCompilers::ASR::ttype_t *,struct std::less<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >,class std::allocator<struct std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const ,struct LCompilers::ASR::ttype_t *> > >,class std::map<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,struct LCompilers::ASR::symbol_t *,struct std::less<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >,class std::allocator<struct std::pair<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const ,struct LCompilers::ASR::symbol_t *> > >,struct LCompilers::SymbolTable *,struct LCompilers::SymbolTable *,struct LCompilers::ASR::Function_t *,struct LCompilers::ASR::Function_t *)" (?pass_instantiate_function_body@LCompilers@@YAXAEAVAllocator@@V?$map@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PEAUttype_t@ASR@LCompilers@@U?$less@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@2@V?$allocator@U?$pair@$$CBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PEAUttype_t@ASR@LCompilers@@@std@@@2@@std@@V?$map@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PEAUsymbol_t@ASR@LCompilers@@U?$less@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@2@V?$allocator@U?$pair@$$CBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PEAUsymbol_t@ASR@LCompilers@@@std@@@2@@4@PEAUSymbolTable@1@3PEAUFunction_t@ASR@1@4@Z) referenced in function "public: void __cdecl LCompilers::LFortran::BodyVisitor::visit_Instantiate(struct LCompilers::LFortran::AST::Instantiate_t const &)" (?visit_Instantiate@BodyVisitor@LFortran@LCompilers@@QEAAXAEBUInstantiate_t@AST@23@@Z)
src\bin\lfortran.exe : fatal error LNK1120: 1 unresolved externals
```

**Comment by certik** [comment]
There is a warning that needs to be fixed:
```
/home/runner/work/lfortran/lfortran/src/libasr/asr.h: In member function ‘void LCompilers::LFortran::SymbolTableVisitor::visit_Instantiate(const LCompilers::LFortran::AST::Instantiate_t&)’:
/home/runner/work/lfortran/lfortran/src/libasr/asr.h:2373:17: error: ‘op’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
 2373 |         n->m_op = a_op;
      |         ~~~~~~~~^~~~~~
/home/runner/work/lfortran/lfortran/src/lfortran/semantics/ast_symboltable_visitor.cpp:2290:32: note: ‘op’ was declared here
 2290 |                 ASR::binopType op;
      |                                ^~
```

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 3]
```diff
@@ -826,6 +829,33 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         }
         restriction_subs[restriction_name] = sym_arg;
     }
+*/
```
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:832]: Let's remove it if it is not needed.

**Comment by certik** [comment]
Otherwise I think this is fine. Once you resolve the issues, ping me and I'll do a final review.

**Comment by ansharlubis** [comment]
@certik I found the cause of the wasm problem. In this PR, the instantiated functions are moved into the symbol table of the subroutine/function that generated them.

If we have this program:
```fortran
module template_add_01_m
    implicit none
    private
    public :: add_t

    requirement R(T, F)
    ...
    end requirement

    template add_t(T, F)
        requires R(T, F)
        private
        public :: add_generic
    contains
        function add_generic(x, y) result(z)
            type(T), intent(in) :: x, y
            type(T) :: z
            z = F(x, y)
        end function
    end template

contains

    real function func_arg_real(x, y) result(z)
        real, intent(in) :: x, y
        z = x + y
    end function

    subroutine test_template()
        instantiate add_t(real, func_arg_real), only: add_real => add_generic
    end subroutine
end module
```

Before this PR, the ASR would look like:
```
Module: template_add_01_m
  Function: func_arg_real
  Subroutine: test_template
  Function: add_real
```
where the instantiated function add_real is on the same level as the subroutine test_template.

But in this PR I modified it into:
```
Module: template_add_01_m
  Function: func_arg_real
  Subroutine: test_template
    Function: add_real
```
where add_real is nested inside test_template.

And the current wasm generation doesn't consider functions inside of other functions, that's why the function add_real wasn't found.

**Comment by ansharlubis** [comment]
In the meantime I will disable all the wasm tests because all of them produce nested functions.

File: tests/errors/template_error_06.f90 [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+contains
+
+end module
+
+program template_struct_02
+use template_struct_02_m
+
+implicit none
+
+end program
```
Comment (certik) [inline tests/errors/template_error_06.f90:33]: This file is never used for testing?

File: tests/reference/asr-template_error_01-b4ca796.stdout [hunk 5]
```diff
@@ -0,0 +1,2136 @@
+(TranslationUnit
```
Comment (certik) [inline tests/reference/asr-template_error_01-b4ca796.stdout:1]: Is this file not giving an error anymore?

**Comment by certik** [comment]
Thanks! I left some minor comments, I'll review it thoroughly soon.

File: tests/reference/asr-template_error_01-b4ca796.stdout (hunk 5 above)
Comment (ansharlubis) [reply tests/reference/asr-template_error_01-b4ca796.stdout:1]: Seems like I added it by mistake while fixing it.

File: tests/errors/template_error_06.f90 (hunk 4 above)
Comment (ansharlubis) [reply tests/errors/template_error_06.f90:33]: Thanks for pointing that out, I've added the reference test for that too.

Comment (certik) [reply tests/errors/template_error_06.f90:33]: I don't see this file being tested in tests.toml.  You have to add it there, and then commit the new reference files, which should show the error message.

Comment (ansharlubis) [reply tests/errors/template_error_06.f90:33]: I've pushed it to my repo but it seemed to take some time before it's reflected in this PR.

File: tests/reference/asr-template_error_06-1cbc8a9.stderr [hunk 6]
```diff
@@ -0,0 +1,5 @@
+semantic error: Type parameter 'g' not required
```
Comment (certik) [inline tests/reference/asr-template_error_06-1cbc8a9.stderr:1]: I wasn't sure what this error message means. Let's write it like this:
```suggestion
semantic error: Type parameter 'g' is not specified in any requirements
```

File: src/lfortran/semantics/ast_common_visitor.h [hunk 7]
```diff
@@ -2468,6 +2461,9 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                 type = ASRUtils::TYPE(ASR::make_CPtr_t(al, loc));
             } else {
                 if (!v) {
+                    if (is_template) {
+                      throw SemanticError("Type parameter '" + derived_type_name + "' not required", loc);
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:2465]: Something like this I think is better:
```suggestion
                      throw SemanticError("Type parameter '" + derived_type_name + "' is not part of any requirement", loc);
```

File: src/libasr/pass/instantiate_template.cpp [hunk 8]
```diff
@@ -427,14 +513,25 @@ class FunctionInstantiator : public ASR::BaseExprStmtDuplicator<FunctionInstanti
 
 };
 
-ASR::symbol_t* pass_instantiate_template(Allocator &al, std::map<std::string, ASR::ttype_t*> subs,
-        std::map<std::string, ASR::symbol_t*> rt_subs, SymbolTable *current_scope,
-        SymbolTable* template_scope, std::string new_func_name, ASR::symbol_t *sym) {
+ASR::symbol_t* pass_instantiate_symbol(Allocator &al,
```
Comment (certik) [inline src/libasr/pass/instantiate_template.cpp:516]: It seems the return value is never used. What is the purpose of returning it?

File: src/libasr/pass/instantiate_template.cpp [hunk 9]
```diff
@@ -427,14 +513,25 @@ class FunctionInstantiator : public ASR::BaseExprStmtDuplicator<FunctionInstanti
...
-    ASR::Function_t* func = ASR::down_cast<ASR::Function_t>(sym2);
-    FunctionInstantiator tf(al, subs, rt_subs, current_scope, template_scope, new_func_name);
-    ASR::asr_t *new_function = tf.instantiate_Function(func);
-    return ASR::down_cast<ASR::symbol_t>(new_function);
+    SymbolInstantiator t(al, type_subs, symbol_subs,
+        current_scope, template_scope, new_sym_name);
+    return t.instantiate_symbol(sym2);
+}
+
+ASR::symbol_t* pass_instantiate_function_body(Allocator &al,
```
Comment (certik) [inline src/libasr/pass/instantiate_template.cpp:527]: The same here.

**Review by certik — APPROVED** [review]
Thanks, I think this looks good. I left some comments, they are all minor. I am going to merge this, and you can fix them up in subsequent PRs.

**Comment by certik** [comment]
The `instantiate_template` is not quite like the other passes --- it accept more arguments and it is used internally in AST->ASR. It is more of a "utility" than a pass, would you agree? I am improving the ASR pass manager and I can't figure out so far how to fit this pass in.

**Comment by ansharlubis** [comment]
Yes, it is only used to generate new function and there is no ASR conversion happening through it.

**Comment by certik** [comment]
@ansharlubis I've opened up a PR with your fixes here: https://github.com/lfortran/lfortran/pull/1939

**Pull request merged**